### PR TITLE
feature - make grovepi easy to install

### PR DESCRIPTION
**Pull request by RobertLucian**
This one has been thoroughly tested. It's backward compatible and that can be seen in the following screenshot (take a look at those options used by default).

grovepi

Apart from that, here's the attached README for installing the repo.

## Installing

You need internet access for the following step(s).

The quickest way for installing the GrovePi is to enter the following command:
```
curl -kL dexterindustries.com/update_grovepi | bash
```

By default, the GrovePi package is installed system-wide and [script_tools](https://github.com/DexterInd/script_tools) is completely updated each time the script is ran.

An example using options appended to the command can be:
```
curl -kL dexterindustries.com/update_grovepi | bash -s --user-local --no-update-aptget --no-dependencies
```

## Command Options

The options that can be appended to this command are:

* `--no-dependencies` - skip installing any dependencies for the GrovePi. It's supposed to be used on each consecutive update after the initial install has gone through.
* `--no-update-aptget` - to skip using `sudo apt-get update` before installing dependencies. For this to be useful, `--no-dependencies` has to be not used.
* `--bypass-pkg-scriptools` - skips installing/updating the python package for  [script_tools](https://github.com/DexterInd/script_tools).
* `--user-local` - install the python package for the GrovePi in the home directory of the user. This doesn't require any special read/write permissions: the actual command used is (`python setup.py install --force --user`).
* `--env-local` - install the python package for the GrovePi within the given environment without elevated privileges: the actual command used is (`python setup.py install --force`).
* `--system-wide` - install the python package for the GrovePi within the sytem-wide environment with `sudo`: the actual command used is (`sudo python setup.py install --force`).

Important to remember is that `--user-local`, `--env-local` and `--system-wide` options are all mutually-exclusive - they cannot be used together.
As a last thing, different versions of it can be pulled by appending a corresponding branch name or tag.

### CODE REVIEW

File: Software/Python/setup.py [hunk 1]
```diff
@@ -44,7 +44,7 @@
 import setuptools
 setuptools.setup(
     name = "grovepi",
-    version = "1.0.1",
+    version = "1.0.0",
```
Comment (CleoQc) [inline Software/Python/setup.py]: why are we going back in versions?

File: Script/install.sh [hunk 2]
```diff
@@ -1,5 +1,4 @@
 #! /bin/bash
-curl --silent https://raw.githubusercontent.com/DexterInd/script_tools/master/install_script_tools.sh | bash
```
Comment (CleoQc) [inline Script/install.sh]: You've removed the curl install_script_tools.sh 
Is it done elsewhere? Cause you can't assume it will be on the SD card if people run this script directly.

File: Script/install.sh [hunk 3]
```diff
@@ -11,115 +10,30 @@ REPO_PATH=$(readlink -f $(dirname $0) | grep -E -o "^(.*?\\GrovePi)")
 
 source $DEXTERSCRIPT/functions_library.sh
```
Comment (CleoQc) [inline Script/install.sh]: Since the curl is removed you can't assume you will be able to source the file

File: Script/install.sh [hunk 4]
```diff
@@ -11,115 +10,30 @@ REPO_PATH=$(readlink -f $(dirname $0) | grep -E -o "^(.*?\\GrovePi)")
...
 
-identify_cie() {
```
Comment (CleoQc) [inline Script/install.sh]: Why are those removed? They are needed for branding.

File: Script/install.sh [hunk 5]
```diff
@@ -11,115 +10,30 @@ REPO_PATH=$(readlink -f $(dirname $0) | grep -E -o "^(.*?\\GrovePi)")
...
+  	feedback "Installing Dependencies"
+  	echo "======================="
+  	sudo apt-get install git libi2c-dev i2c-tools arduino minicom -y
+    sudo apt-get purge python-rpi.gpio -y
+  	sudo apt-get purge python3-rpi.gpio -y
+    sudo apt-get install python-pip python-smbus python-dev python-serial python-rpi.gpio python-scipy -y
+    sudo apt-get install python3-smbus python3-dev python3-rpi.gpio python3-scipy -y
+  	sudo pip install -U RPi.GPIO
+    sudo pip3 install -U RPi.GPIO
+    sudo pip2 install numpy
```
Comment (CleoQc) [inline Script/install.sh]: Does numpy really have to be installed for GrovePi to work? Can't we just install it for the projects that actually need it?

File: Script/update_grovepi.sh [hunk 6]
```diff
@@ -5,34 +5,147 @@ PIHOME=/home/pi
...
+  esac
+done
 
-if [ -d "$GROVEPI_DIR" ]; then
-    echo "GrovePi Directory Exists"
-    cd $GROVEPI_DIR             # Go to directory
-    sudo git fetch origin       # Hard reset the git files
-    sudo git reset --hard  
-    sudo git merge origin/master
+# show some feedback for the GrovePi
```
Comment (CleoQc) [inline Script/update_grovepi.sh]: Please keep this branding in a function so it can be out of sight, out of mind when we work in the script.

File: Script/update_grovepi.sh [hunk 7]
```diff
@@ -5,34 +5,147 @@ PIHOME=/home/pi
...
+echo " / ____|                  |  __ (_)  "
+echo "| |  __ _ __ _____   _____| |__) |   "
+echo "| | |_ | '__/ _ \ \ / / _ \  ___/ |  "
+echo "| |__| | | | (_) \ V /  __/ |   | |  "
+echo " \_____|_|  \___/ \_/ \___|_|   |_|  "
+echo " "
+
+# show some feedback on the console
+if [ -f $DEXTERSCRIPT/functions_library.sh ]; then
+  source $DEXTERSCRIPT/functions_library.sh
```
Comment (CleoQc) [inline Script/update_grovepi.sh]: This is silly, either functions_library.sh is used or it's not.   Pull it in, curl it if necessary, and make use of it.

File: Script/update_grovepi.sh [hunk 8]
```diff
@@ -5,34 +5,147 @@ PIHOME=/home/pi
...
+  feedback "Welcome to GrovePi Installer."
 else
-	echo "Cloning"
-    cd $PIHOME/$DEXTER/
-    git clone https://github.com/DexterInd/GrovePi
-    cd GrovePi
+  echo "Welcome to GrovePi Installer."
+fi
+echo "Updating GrovePi for $selectedbranch branch with the following options:"
```
Comment (CleoQc) [inline Script/update_grovepi.sh]: these echo could totally be feedback instead

File: Script/update_grovepi.sh [hunk 9]
```diff
@@ -5,34 +5,147 @@ PIHOME=/home/pi
...
+optionslist+=("$selectedbranch")
+[[ $usepython3exec = "true" ]] && optionslist+=("--use-python3-exe-too")
+[[ $updaterepo = "true" ]] && optionslist+=("--update-aptget")
+[[ $installdependencies = "true" ]] && optionslist+=("--install-deb-deps")
+[[ $install_pkg_scriptools = "true" ]] && optionslist+=("--install-python-package")
+
+echo "Options used for script_tools script: \"${optionslist[@]}\""
+
+# update script_tools first
+curl --silent -kL dexterindustries.com/update_tools > $PIHOME/tmp_script_tools.sh
```
Comment (CleoQc) [inline Script/update_grovepi.sh]: why in a temporary spot??  Just curl it once and for all.

File: Script/update_grovepi.sh [hunk 10]
```diff
@@ -5,34 +5,147 @@ PIHOME=/home/pi
...
-feedback "--> Start GrovePi update install."
-feedback "---------------------------------"
-pushd $PIHOME/$DEXTER/GrovePi/Script > /dev/null
+# HAVE TO UNCOMMENT ONCE check_internet makes it to script_tools
+# check if there's internet access,
+# otherwise straight out exit the script
+# check_internet
+
+# needs to be sourced from here when we call this as a standalone
+source /home/pi/$DEXTER/lib/$DEXTER/script_tools/functions_library.sh
```
Comment (CleoQc) [inline Script/update_grovepi.sh]: Best practice is always to curl at the top and source it at the top, instead of having to scroll down hundred lines to see if it's available or not.

That way, it's standard from one script to the next.

File: Script/update_grovepi.sh [hunk 11]
```diff
@@ -5,34 +5,147 @@ PIHOME=/home/pi
...
+
+# create folders recursively if they don't exist already
+mkdir -p $DEXTER_PATH
```
Comment (CleoQc) [inline Script/update_grovepi.sh]: oh, I had never noticed that -p option before. Nice to have!

File: Script/update_grovepi.sh (hunk 9 above)
Comment (RobertLucian) [reply Script/update_grovepi.sh]: If you read the source code, you'll see that's not possible, because `${optionslist[@]}` has to be evaluated first and then they can be used. Since it gets evaluated you end up with something that doesn't look like a command.

Comment (CleoQc) [reply Script/update_grovepi.sh]: The curl itself doesn't make use of optionlist. And only the curl is needed to be able to source functions_library.
So you can do it at the top. 
curl
source functions_library
then later on, run the install script.

And I still don't get why it's needed in a tmp folder.

File: Script/install.sh (hunk 2 above)
Comment (RobertLucian) [reply Script/install.sh]: Since `install.sh` is such a mess, it shouldn't be used for installing the GrovePi. Therefore, installing the GrovePi only works by using the command for updating the repo. All after, there's a need for internet access in both cases. 

I have moved this out of here, because it now resides in `update_grovepi.sh`. It's not okay to call it twice in an update - `update_grovepi.sh` is used when we do the curl stuff.

File: Script/install.sh (hunk 4 above)
Comment (RobertLucian) [reply Script/install.sh]: There are put back in `update_grovepi.sh`. They are supposed to be there - otherwise you'll get a message with "Welcome to GrovePi's installer" somewhere in the middle on the installation or none at all.

File: Script/install.sh (hunk 5 above)
Comment (RobertLucian) [reply Script/install.sh]: Yes. It's needed. We already have it installed for 8 months or so.

Comment (RobertLucian) [reply Script/install.sh]: Unless .. unless we re-write a part of the code to that's processed with numpy. It should be doable.

File: Script/update_grovepi.sh (hunk 6 above)
Comment (RobertLucian) [reply Script/update_grovepi.sh]: I can put it in a file separate other than this one and the `install.sh` script.

File: Script/update_grovepi.sh (hunk 7 above)
Comment (RobertLucian) [reply Script/update_grovepi.sh]: Since `install.sh` is a big mess, I had to containerize it and make it work like a module that only installs dependencies and I moved the rest here.
Above that, we have to first parse the arguments and right after that we can curl the command. We can't curl it twice for this matter, it doesn't make sense. We can't curl it and then parse the arguments. So, instead we check if, `script_tools` is installed, and if it is, we use the `feedback` function, else we just use a regular echo. This means that `feedback` will work only if a certain robot has already been installed once, and that's fine.

File: Script/update_grovepi.sh (hunk 8 above)
Comment (RobertLucian) [reply Script/update_grovepi.sh]: See the other comment you left:
> This is silly, either functions_library.sh is used or it's not. Pull it in, curl it if necessary, and make use of it.

File: Script/update_grovepi.sh (hunk 7 above)
Comment (CleoQc) [reply Script/update_grovepi.sh]: install.sh is what is referenced in tutorials all over the place as being the proper way to install GrovePi. 
We have to live with that. You can't just go in and move stuff from one script to another.

File: Script/update_grovepi.sh (hunk 10 above)
Comment (RobertLucian) [reply Script/update_grovepi.sh]: While I agree with you, the thing is, you don't know if `script_tools` is installed so you have got to source after you curl it. Regarding the curling stuff, well you can't curl it until you have a list of options to pass it to it - which means you have to parse the input - which takes plentiful of slocs.

Unfortunately, this is the material we have to work with - if it would have been to have a new installing mechanism, it would be ages ahead and it would always follow the best practices. But we are here to solve the current issue without being too concerned whether this the best version or not - we are only focusing on making it work asap. For instance, there's a reason I kept the old `install.sh`.

File: Script/update_grovepi.sh (hunk 11 above)
Comment (RobertLucian) [reply Script/update_grovepi.sh]: Yeah, it trims down the code a lot. 😀

File: Software/Python/setup.py (hunk 1 above)
Comment (RobertLucian) [reply Software/Python/setup.py]: Well, we don't really do any versioning here so I thought "why not have a round number". What do you think?

I don't have any strong preference here.

File: Script/update_grovepi.sh (hunk 9 above)
Comment (RobertLucian) [reply Script/update_grovepi.sh]: I could move the curl command up top and save the temp file. But do we want to do that? It's like the temp file is exist for a bit more time.

On the other hand, the curl command *no longer makes use of optionlist* because we're using the temp file. Now, when passing `${optionslist[@]}` to the curl command through piping, `{optionslist[@]}` gets evaluated and returns exits due to not being an actual command, so that's the reason I used a temp file. 
There may be a way to bypass the evaluation of the variable, but that could take more time to research and instead, I would want to have this out ASAP.

File: Script/update_grovepi.sh (hunk 7 above)
Comment (RobertLucian) [reply Script/update_grovepi.sh]: I can't contest the presence of `install.sh` in other tutorials.

How do you envision the solution?

File: Software/Python/setup.py (hunk 1 above)
Comment (CleoQc) [reply Software/Python/setup.py]: We already have a 1.0.0 out there and a 1.0.1, so no going back.
In fact I think the version on pip is 1.0.0 (I may be wrong, it's been years!)

**Comment by RobertLucian** [comment]
@mattallen37 I have modified this PR so that it performs the same way as https://github.com/DexterInd/GoPiGo/pull/307. You might also want to test this.

**Comment by RobertLucian** [comment]
1. Quiet mode for text
2. Functions for every step of the install script
3. Possibly use another duplicate of feedback
4. Possibly remove wiringpi
5. Move from first person to 3rd person logs
6. Add `echo "Installing GrovePi dependencies and package. This might take a while.."` on a needed basis.
7. Put the install.sh and everything else in just on if statement
8. Add nodejs support
